### DISCUSSION TITLE
JSON logging format

### DISCUSSION THREAD

**Discussion by jangaraj** [keycloak/keycloak#10209] [Archive]
Wildfly based Keycloak provides option to change logging format to JSON, e.g.:
```
embed-server --server-config=standalone-ha.xml --std-out=echo
/subsystem=logging/json-formatter=JSON/:add
/subsystem=logging/console-handler=CONSOLE/:write-attribute(name=named-formatter,value=JSON)
stop-embedded-server
```
Quarkus based Keycloak doesn't have this option (v 17.0.0). generated logs are in the text format only. [Logging options](https://www.keycloak.org/server/logging):
```
--log-format <format>
                     The format of log entries. If the format has spaces in it, you need to escape
                       the value such as "<format>". Default: %d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c]
                       (%t) %s%e%n.
--log-level <category:level>
                     The log level of the root category or a comma-separated list of individual
                       categories and their levels. For the root category, you don't need to
                       specify a category. Default: info.
```
Naive solution with current config option:
```
log-format={"time":"%d{yyyy-MM-dd HH:mm:ss,SSS}", "category":"%c", "thread-name":"%t", "thread-id":"%t{id}", "process-id":"%i", "message":"%m", "exception":"%e"}%n
```
doesn't work very well - it may generate empty fields, (e.g. empty exception, when there is no exception).

1.  It will help if Keycloak admin will have easy option switch from the text log format to json format (which can be then parsed easily with standard logging tools). This probably can be easily implemented via quarkus extension `quarkus-logging-json`.

2. It will help if also full message (`%m`)/short message (`%s`) is generated in the JSON format. Use case: Keycloak event messages don't need additional parsing - default Keycloak JSON logs will be enough to generate/calculate/aggregate all possible metrics/reports, e.g. succesfull/failed logins of (unique) users/clients per selected time period, ...

Example of Keycloak event in the text format: 
```
type=CLIENT_LOGIN, realmId=XYZ, clientId=CLIENTID, userId=USERID, ipAddress=X.X.X.X, token_id=TOKENID, grant_type=client_credentials, scope='email profile', client_auth_method=client-secret, username=USERNAME, authSessionParentId=PARENTID, authSessionTabId=TABID
```
Which should be logged in the JSON format as:
```
{
	"type": "CLIENT_LOGIN",
	"realmId": "XYZ",
	"clientId": "CLIENTID",
	"userId": "UERID",
	"ipAddress": "X.X.X.X",
	"token_id": "TOKENID",
	"grant_type": "client_credentials",
	"scope": "'email profile'",
	"client_auth_method": "client-secret",
	"username": "USERNAME",
	"authSessionParentId": "PARENTID",
	"authSessionTabId": "TABID"
}
```

@dasniko [mentioned that in the community](https://keycloak.discourse.group/t/logging-in-full-json-format/7179/3):
> This is not possible with the default logging event listener, as it is concatenating the message to a string. The actual message is the value to the key message.
>
>If you also want to have the message details printed as JSON output, you‘ll have to implement a custom event listener, doing exactly this.

**Comment by DGuhr** [comment]
Hey. On point 1 i am totally aligned and will provide a PR shortly. Can you elaborate a bit more on point 2? What option would you set and what would be your expected outcome/log output? Thanks!

**Reply by jangaraj** [reply]
Description updated.

**Reply by dasniko** [reply]
Be careful not to mix up two different things.

One is the general logging format of log messages. That should be easily switchable to a JSON format instead of plain text.

Second is the raw message payload from the default events logger. As the message payload is concatenated as a string in the events listener, this would mean to change it there. And as it is "just" the value of the `message` key in the actual log entry, it would be stringified JSON object as value in the JSON-ified log message.

**Comment by DGuhr** [comment]
Initial json console logging support is now in latest main, see #10576 - expect some more options when we get to #10618 - we need quarkus 2.8.0 to be released though.
